nbsite/nbbuild.py

```diff
@@ -178,34 +187,170 @@ class FixNotebookLinks(Preprocessor):
+        Link target: "..foo/b.ipynb"
+
+        String "foo [a](b.ipynb) bar [c](..foo/d.ipynb#spam)" would return
+        iterable, which when converted to a list would be: [
+             [('[a](b.ipynb)', 'b.ipynb')]
+             [('[c](../foo/d.ipynb#spam)', '../foo/d.ipynb')]
+        ]
+
+        """
+        for match in re.finditer(r"(\[.+?\]\((.+?\.ipynb)#*?.*?\))", markdown_text):
+            yield match.groups()
+
+    @classmethod
+    def _iter_source_file_candidates(cls, nb_filepath: str) -> Iterable[str]:
+        """Gets potential link targets corresponding to a notebook file path.
+        Potential link targets (source files paths) are formed by using the
+        relative path of the link target notebook, and adding the file
+        extensions from `cls.file_types`. In addition if the the filename part
+        of the `nb_filepath` starts with digits + one of ('-', '_', ' '),
+        the stem without the number prefix is used to form a potential link
+        target.
+
+        Parameters
+        ----------
+        nb_filepath:
+            The filepath of a link target a notebook. Must end with ".ipynb".
+            The links are relative to the directory of the notebook which
+            contains the markdown link. Example: "../foo/10-Some.ipynb"
+
+        Example
+        -------
+        >>> list(_iter_source_file_candidates("../foo/01-Some_Notebook.ipynb"))
+        [
+            "../foo/01-Some_Notebook.rst",
+            "../foo/01-Some_Notebook.md",
+            "../foo/Some_Notebook.rst",
+            "../foo/Some_Notebook.md",
+        ],
+        """
+
+        if not nb_filepath.endswith('.ipynb'):
+            raise ValueError('nb_filepath must end with .ipynb')
+
+        nb_path_without_extension = nb_filepath[:-6]
+        for extension in cls.file_types:
+            yield f"{nb_path_without_extension}.{extension}"
+
+        for extension in cls.file_types:
+            *directory, stem = re.split('(/)', nb_path_without_extension)
+            match = re.match(r"\d+[ -_](.*)", stem)
+
+            if match:
+                yield ''.join(directory) + match.group(1) + '.' + extension
+
+
+    @staticmethod
+    def _create_target_link(nb_link: str, target_relpath:str) -> str:
+        """Using a markdown link `nb_link`, create a markdown link which
+        changes the link target file to be `target_relpath`.
+
+        Example
+        -------
+        If nb_link is "[sometext](../../somepath/01-subject.ipynb#spam)", and
+        the target_filename is "../otherpath/subject.rst", returns
+        "[sometext](../otherpath/subject.rst#spam)"
+
+        Notes:
+        * If you pass as nb_link text which contains no markdown links, a
+          ValueError will be raised
+        * If you pass as nb_link text which contains more than one markdown
+          links, the behaviour is undefined.
+        """
+
+        # Example: [sometext](../../somepath/subject.ipynb#spam)
+        match = re.match(
```

Comment:
Is this the same regex as you are using in `extract_link`? If so can't we pull this out of the class and have it as a variable with `re.compile`.

Reply:
Good question. I had to go back and look at the code. They're very similar, but since they're capturing different groups for different purposes, I'm not sure if it's possible to combine them. The best way to check what they're doing is in the unit tests. 

The first one, in the extract_links, will match two types of things:
- The full markdown link, like `"[a](..some/book.ipynb#spam)"`
- Just the relative path of the notebook, like:  `"..some/book.ipynb"`

The second one, in the create_target_link, will match four things. They all are used to make sure this is indeed a markdown link, and two of them (first and last) are used when creating the new link (to the .rst / .md file):
-  `"[a]("`  *used as first part of the new link*
- `"..some/book"`  *only used for checking*
- `".ipynb"` *only used for checking. Could perhaps be in the same group as the previous*
- `"#spam)"` *used as the ending part of the new link*

The new link could be then `"[a](..any/source.rst#spam)"`. Now after I wrote this it looks like it would be possible to use the second regex also in the extract links part. I'll try that!

Reply:
Got it working with single regex pattern. Good suggestion, thanks!